### ISSUE TITLE
Support PEP 660 editable packages.

### ISSUE THREAD

**Issue by alanhdu** [microsoft/pyright#3846]
**Describe the bug**

I have a package that is configured via `pyproject.toml` and install it in editable mode for local development (as specified in PEP660). When I do this though, `pyright` complains about being unable to resolve the imports:

**To Reproduce**

Create a directory with the following files:

```bash
mkdir foopkg
echo "CONSTANT = 3" > foopkg/__init__.py
echo "
[build-system]
requires = ['setuptools']
build-backend = 'setuptools.build_meta'

[project]
name = 'foopkg'
" > pyproject.toml
pip install -e .

```

This creates a simple `foopkg` package that can be editably installed (see [PEP 660](https://peps.python.org/pep-0660/)).  if you now create a simple `scratch.py` file with:
```python
import foopkg
print(foopkg.CONSTANT)
```
This will run fine in the Python interpreter but `pyright` will throw up `error: Import "foopkg" could not be resolved (reportMissingImports)`.

**Expected behavior**

I would expect pyright to know how to automatically resolve this editable package, like it does if we don't use PEP 660 and instead use `setup.py` to define the package installation.

**VS Code extension or command-line**
I've been running `pyright` via the command-line (`1.1.267`).

**Comment by erictraut** [comment]
We currently have no plans to support PEP 660. The approach that it uses (a dynamic loader hook) is highly problematic for static analysis tools. There has been significant discussion about this in the [pylance issue tracker](https://github.com/microsoft/pylance-release/issues/2114) and in [the typing-sig](https://mail.python.org/archives/list/typing-sig@python.org/thread/IIVBPYDZR5T5BGPAWFVYS5ZPYDXGVHQN/#MK424OLZOGCKW2X7STERHJJRRNCGE73M).

If you use PEP 660 or other techniques that employ dynamic import hooks, you'll need to manually configure the import paths using the `extraPaths` configuration.

**Comment by hauntsaninja** [comment]
Note that an editable installation based on a static entry in `.pth` is valid under PEP 660 and is supported by all type checkers and IDEs.

Despite consensus amongst authors of type checkers and IDEs that support for import hooks isn't feasible, setuptools went ahead and changed their default method of installation in v64. See https://github.com/pypa/setuptools/issues/3518

**Comment by shaoran** [comment]
@erictraut If I understand your last comment correctly, pyright will never support PEP 660, right?

**Comment by erictraut** [comment]
@shaoran, PEP 660 covers a number of pieces of functionality. We specifically have no plans to support dynamic loader hooks in pyright, and it's unlikely that other static analysis tools will either. Static analysis tools require a declarative mechanism for identifying import paths.

**Comment by shaoran** [comment]
@erictraut ok, thanks for the explanation.

if I put a `pyrightconfig.json` in my HOME directory and specify  `extraPaths` there with the path of my library, then pyright should be able to resolve my library independent of my current working directory, right? What would happen if a project of mine already has a `pyrightconfig.json` in the project's root directory. Does pyright stop looking for config files when it found one?

**Comment by erictraut** [comment]
@shaoran, by default pyright looks for a `pyrightconfig.json` in your project's root directory. It supports "chaining" of config files via "extends". See [this documentation](https://microsoft.github.io/pyright/#/configuration?id=environment-options) for details.
